### DISCUSSION TITLE
Add rectangle through odb scripting

### DISCUSSION THREAD

**Discussion by OuDret** [The-OpenROAD-Project/OpenROAD#5087] [Q&A]
Hello,

I was trying to place some rectangles/boxes using odb scripting but I did not manage to make it work. 

Is it even possible?

**Comment by maliberty** [comment]
What sort of rectange/box are you trying to add?  The odb api from db.h is swig'ed to tcl.

**Reply by OuDret** [reply]
I wanted to do custom pdn ring but I'm not achieving it at all. 

I tried adding dbWires to VDD dbNet but function to add new segments is private. I have also tried to create directly dbBox & dbFill but they have to be contained on dbInstance so it was not a good idea aswell.

What approach would you recommend?

Is there any kind of db visualizer for odb? It would be really helpful to understand db structure alongside db.h classes and comments. 

Thank you

**Reply by maliberty** [reply]
What are you not able to do with the current pdngen that requires custom code?

`openroad -gui` lets you load a design and view it.

**Reply by OuDret** [reply]
I am using following pdn for which placement and routing works fine:
```
add_global_connection -net {VDD} -inst_pattern {.*} -pin_pattern {^VDD$} -power
add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {^VSS$} -ground
add_global_connection -net {VBG} -inst_pattern {.*} -pin_pattern {^VBG$} -power

global_connect

set_voltage_domain -name {CORE} -power {VDD} -ground {VSS} -secondary_power {VBG}
define_pdn_grid -name {grid} -voltage_domains {CORE} 

add_pdn_stripe -grid {grid} -width {300} -layer {metal2} -followpins
add_pdn_stripe -grid {grid} -layer {metal1} -pitch {2900} -spacing {0} -offset {1265} -width {50} -nets VBG -extend_to_boundary
```
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/74424277/fb6ac475-2dc1-48f3-ad9d-f1673581c6dd)
I would like it left and right core edges to be like this:
![img1](https://github.com/The-OpenROAD-Project/OpenROAD/assets/74424277/3d21790c-e02b-4b54-93b4-86bf4972d692)
![img2](https://github.com/The-OpenROAD-Project/OpenROAD/assets/74424277/04dc4c3c-1b8c-4c84-8f50-147d3b51c1e4)
Which is not exactly what pdn ring do.

Right now I do through Klayout pya but then I can not use left and right edges for pin placement, since it will short my external add-ons. What approach would you recommend me to take? The plan is to get QFN64 die... 

PS:
By database viewer I meant something like this:
![img4](https://github.com/The-OpenROAD-Project/OpenROAD/assets/74424277/53113f33-3b8a-4d38-a83b-5001f69cbf7f)

**Comment by maliberty** [comment]
> Which is not exactly what pdn ring do.

What do you get with a ring?

**Reply by OuDret** [reply]
I got this. Which is not what I want since I need one of the power nets not only to be on different layer but also this layer must be the same for horitzontal and vertical stripes.

![Screenshot from 2024-05-24 07-54-02](https://github.com/The-OpenROAD-Project/OpenROAD/assets/74424277/f6d559c4-8e6a-4692-b6ff-16057b245441)

**Reply by maliberty** [reply]
@gadfort what do you think about taking a single layer for add_pdn_ring ?

**Reply by gadfort** [reply]
I think it would be fine, we was discussed with Seyi earlier, but it turned out to not be required at that time.
It would just require some updates to the error checking I think (nothing in the ring code should care that there are two different layers). Connecting with vias on the non-default direction would the only thing I would worry about.

**Reply by maliberty** [reply]
@gadfort is this something you have time for or should I suggest odb calls?

**Reply by gadfort** [reply]
I have time this weekend to take a look at it
